From 40a7c022423ecac60fb1289b6cf746aad4aa61f3 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Thu, 7 Nov 2019 19:59:41 -0800
Subject: [PATCH 0683/1851] v4l: xilinx: axis-switch: sink and source pad
 should have same value in TDEST routing mode

commit bd0a079d7a663fe91d6df11c106c11f70a36af1f from
https://github.com/Xilinx/linux-xlnx.git

When there is only one sink port and TDEST routing mode is present, the
source pads should have same format as sink pad.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../platform/xilinx/xilinx-axis-switch.c      | 22 +++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-axis-switch.c b/drivers/media/platform/xilinx/xilinx-axis-switch.c
index 3963e364570a..86cd8585116d 100644
--- a/drivers/media/platform/xilinx/xilinx-axis-switch.c
+++ b/drivers/media/platform/xilinx/xilinx-axis-switch.c
@@ -187,6 +187,28 @@ static int xvsw_set_format(struct v4l2_subdev *subdev,
 		return xvsw_get_format(subdev, cfg, fmt);
 	}
 
+	if (xvsw->nsinks == 1 && fmt->pad != 0) {
+		struct v4l2_mbus_framefmt *sinkformat;
+
+		/*
+		 * in tdest routing if there is only one sink then all the
+		 * source pads will have same property as sink pad, assuming
+		 * streams going to each source pad will have same
+		 * properties.
+		 */
+
+		/* get sink pad format */
+		sinkformat = xvsw_get_pad_format(xvsw, cfg, 0, fmt->which);
+
+		fmt->format = *sinkformat;
+
+		/* set sink pad format on source pad */
+		format = xvsw_get_pad_format(xvsw, cfg, fmt->pad, fmt->which);
+		*format = *sinkformat;
+
+		return 0;
+	}
+
 	/*
 	 * In TDEST routing mode, one can set any format on the pad as
 	 * it can't be checked which pad's data will travel to
-- 
2.31.1

